src/main/java/org/osc/manager/ism/api/IsmDeviceApi.java

```diff
@@ -138,80 +189,184 @@ public Void call() throws Exception {
+
             @Override
-            public ManagerDeviceMemberElement call() throws Exception {
-                CriteriaBuilder cb = em.getCriteriaBuilder();
-                
-                CriteriaQuery<DeviceMember> q = cb.createQuery(DeviceMember.class);
-                Root<DeviceMember> r = q.from(DeviceMember.class);
-                q.select(r).where(cb.and(cb.equal(r.get("name"), name), cb.equal(
-                        r.get("parent").get("id"), vs.getId())));
-                
-                DeviceMember result = em.createQuery(q).getSingleResult();
-                return result == null ? null : new MemberDevice(result.getId(), result.getName());
+            public DeviceMemberEntity call() throws Exception {
+                CriteriaBuilder criteriaBuilder = IsmDeviceApi.this.em.getCriteriaBuilder();
+
+                CriteriaQuery<DeviceMemberEntity> query = criteriaBuilder.createQuery(DeviceMemberEntity.class);
+                Root<DeviceMemberEntity> r = query.from(DeviceMemberEntity.class);
+                query.select(r).where(criteriaBuilder.and(criteriaBuilder.equal(r.get("name"), name)));
+                List<DeviceMemberEntity> result = IsmDeviceApi.this.em.createQuery(query).getResultList();
+                return result.isEmpty() == true ? null : result.get(0);
             }
         });
     }
 
     @Override
     public List<? extends ManagerDeviceMemberElement> listDeviceMembers() throws Exception {
 
-        return txControl.supports(new Callable<List<? extends ManagerDeviceMemberElement>>() {
-            
+        return listDeviceMembers(null);
+    }
+
+    public List<? extends ManagerDeviceMemberElement> listDeviceMembers(Long parentId) throws Exception {
+
+        return this.txControl.supports(new Callable<List<DeviceMemberEntity>>() {
+
             @Override
-            public List<? extends ManagerDeviceMemberElement> call() throws Exception {
-                CriteriaBuilder cb = em.getCriteriaBuilder();
-                
-                CriteriaQuery<DeviceMember> q = cb.createQuery(DeviceMember.class);
-                Root<DeviceMember> r = q.from(DeviceMember.class);
-                q.select(r).where(cb.equal(r.get("parent").get("id"), vs.getId()));
-                
-                List<MemberDevice> devices = new ArrayList<MemberDevice>();
-                for (DeviceMember dev : em.createQuery(q).getResultList()) {
-                    devices.add(new MemberDevice(dev.getId(), dev.getName()));
+            public List<DeviceMemberEntity> call() throws Exception {
+
+                CriteriaBuilder criteriaBuilder = IsmDeviceApi.this.em.getCriteriaBuilder();
+                CriteriaQuery<DeviceMemberEntity> query = criteriaBuilder.createQuery(DeviceMemberEntity.class);
+                Root<DeviceMemberEntity> r = query.from(DeviceMemberEntity.class);
+                if (parentId != null) {
+                    query.select(r)
```

Comment:
See here you are using the parent. Now lets say someone calls this and gets a list of device members with their ids for a given parent, i.e.: parent id: 1, parent vsid: 2. Now i try to delete or retrieve or update one of those device members with a parent id: 2 or a parentvsid: 2. Your apis will succeed because the device member was found by id (no parent look up) when it should fail since i am providing the wrong parentid (from the rest api)  or parent vsid ( from the sdk).